src/core/parse/core/parse_worker.hpp

```diff
@@ -0,0 +1,273 @@
+		{}
+
+		bool proc(const token_t& t, dest_t& f);
+
+		ParseWorker& rewind() { handler().reset_(); handler_i_ = 0; return *this; }
+		ParseWorker& reset() { itered_ = 0; return rewind(); }	
+		ParseWorker& block() { blocker_ = true; return *this; }
+		ParseWorker& limit(size_t iters) { iters_ = iters; return *this; }
+		ParseWorker& working_at(size_t i) { working_at_ = i; return *this; }
+		handler_t& handler_at(size_t i) { return handlers_[i % handlers_.size()]; }
+		handler_t& handler(size_t offset) { return handler_at(handler_i_ + offset); }
+		handler_t& handler() { return handler(0); }
+		handler_t& handler_next() { return handler(1); }
+		void stall() { stalling_ = true; }
+
+	protected:
+		handler_arr_t handlers_;
+
+	private:
+		size_t iters_ = INF_ITERS;
+		size_t working_at_ = 1;
+		bool blocker_ = false;
+
+		unsigned int handler_i_ = 0;
+		size_t itered_ = 0;
+		bool stalling_ = false;
+
+		static constexpr size_t INF_ITERS = 0;
+
+		void restart_() { ++itered_; rewind(); }
+		bool indefinite_() const { return iters_ == INF_ITERS; }
+		bool finished_() const { return !indefinite_() && itered_ >= iters_; }
+		bool done_() const { return handler_i_ == handlers_.size(); }
+		bool working_() const { return handler_i_ >= working_at_ && !finished_(); }
+};
+
+/*
+ * Processes based on the current handler.
+ * Returns true on blocking.
+ */
+template <class TokenType, class DestType>
+inline bool ParseWorker<TokenType, DestType>::proc(const token_t& t, dest_t& d)
+{
+	if (finished_()) {
+		return false;
+	}
+
+	if (handler().proc_workers_(t, d)) {
+		return true;
+	}
+
+	if (handler().match(t)) {	
+		if (handler().on_match_) {
+			handler().on_match_(this, t, d);
+		}
+
+		if (handler().done() && !stalling_) {
+			handler().reset_();
+
+			++handler_i_;
+
+			if (done_()) {
+				restart_();
+			}
+		}
+		stalling_ = false;
+		return blocker_;
+	}
+	else if (handler().time()) {
+		restart_();
+	}
+
+	return false;
+}
+
+/*
+ * Match to a symbol based on whether or not it is contained within tokens_;
+ * if neg_ is true, this will match when the symbol is not contained.
+ */
+template <class TokenType, class DestType>
+inline bool ParseWorker<TokenType, DestType>::TokenHandler::match(const token_t& t)
+{
+	if (!neg_ ? tokens_.find(t) != tokens_.end() : tokens_.find(t) == tokens_.end()) {
+		++match_counter_;
+		return true;
+	}
+	++nomatch_counter_;
+	return false;
+}
+
+/*
+ * Process all contained ParseWorker objects when none are "working";
+ * if one is, process only that one until it is done.
+ * Returns true if any workers are blocking.
+ */
+template <class TokenType, class DestType>
+inline bool ParseWorker<TokenType, DestType>::TokenHandler::proc_workers_(const token_t& t, dest_t& d)
+{
+	bool blocked = false;
+
+	if (working_) {
+		blocked = working_->proc(t, d);
+		if (!working_->working_()) {
+			for (worker_t& w : workers_) {
+				w.rewind();	
+			}
+			working_ = nullptr;
+		}
+	}
+	if (!working_) {
```

Comment:
`else`?

Reply:
You'd think! But no, it's important that a child can begin working on the same token the last one stopped. The state of `working_` is able to change within the former conditional.

Reply:
put a comment then xD